File: data-structure/c-cpp/c/DoubleLinkedList/DummyHeadDoubleLinkedList/src/DummyHeadDoubleLInkedLIst.c
#include "DummyHeadDoubleLInkedLIst.h"
#include <stdlib.h>

void init(List* list) {
    Node* dummy_node = (Node*)malloc(sizeof(Node));
    dummy_node -> next = NULL;
    dummy_node -> prev = NULL;
    list -> numOfData = 0;
    list -> head = dummy_node;
}

void insert(List* list, Data data) {
    Node* new_node = (Node*)malloc(sizeof(Node));
    new_node -> data = data;
    new_node -> next = list -> head -> next;
    new_node -> prev = list -> head;
    list -> head -> next = new_node;
}

int first(List* list, Data* data) {
    if (list -> numOfData == 0) return FALSE;
    list -> cur = list -> head -> next;
    *data = list -> cur -> data;
    return TRUE;
}
int next(List* list, Data* data) {
    if (list -> cur -> next == NULL) return FALSE;
    list -> cur = list -> cur -> next;
    *data = list -> cur -> data;
    return TRUE;
}

Data remove(List* list) {
    Node* r_pos = list -> cur;
    Data r_data = r_pos -> data;
    list -> cur -> prev -> next = list -> cur -> next;
    if (list -> numOfData > 1) {
        list -> cur -> next -> prev = list -> cur -> prev;
    }
    list -> numOfData--;
    free(r_pos);
    return r_data;
}
int count(List* list) {
    return list -> numOfData;
}

File: data-structure/java/src/linkedlist/doublelinkedlist/head/dummy/DummyHeadDoubleLinkedLIst.java
package linkedlist.doublelinkedlist.head.dummy;

import exception.NoElementException;
import linkedlist.doublelinkedlist.Node;

public class DummyHeadDoubleLinkedLIst <T>{
    private Node<T> head;
    private Node<T> cur;
    private int numOfElements;

    public DummyHeadDoubleLinkedLIst(){
        this.head = new Node<>(null);
        this.cur = null;
        this.numOfElements = 0;
    }

    public void insert(T data){
        Node<T> newNode = new Node<>(data);
        newNode.next = this.head;
        this.head = newNode;
        this.numOfElements++;
    }
    public T first(){
        if(this.head.next == null) throw new NoElementException();
        this.cur = this.head.next;
        return this.cur.data;
    }
    public T next(){
        if(this.cur.next == null) throw new NoElementException();
        this.cur = this.cur.next;
        return this.cur.data;
    }
    public T remove(){
        Node<T> rPos = this.cur;
        T rData = rPos.data;
        this.cur.before.next = this.cur.next;
        this.numOfElements--;
        return rData;
    }
    public int count(){
        return this.numOfElements;
    }
}
